test/ai/TestGenerateRandomMove.java

```diff
@@ -0,0 +1,349 @@
+		pieces.put(new Coordinate(3, 3), new Rabbit(Owner.Player2));
+		pieces.put(new Coordinate(4, 4), new Dog(Owner.Player2));
+		pieces.put(new Coordinate(5, 5), new Camel(Owner.Player2));
+		pieces.put(new Coordinate(6, 6), new Elephant(Owner.Player2));
+		pieces.put(new Coordinate(7, 7), new Horse(Owner.Player2));
+		this.normalAi = new Ai(Owner.Player2, new Game(new BoardState(pieces)));
+
+		HashMap<Coordinate, AbstractPiece> catLoverPieces = new HashMap<Coordinate, AbstractPiece>();
+		catLoverPieces.put(new Coordinate(0, 0), new Rabbit(Owner.Player1));
+		catLoverPieces.put(new Coordinate(0, 1), new Dog(Owner.Player1));
+		catLoverPieces.put(new Coordinate(1, 0), new Rabbit(Owner.Player1));
+		catLoverPieces.put(new Coordinate(1, 1), new Dog(Owner.Player1));
+
+		catLoverPieces.put(new Coordinate(6, 7), new Cat(Owner.Player2));
+		catLoverPieces.put(new Coordinate(2, 2), new Cat(Owner.Player2));
+		catLoverPieces.put(new Coordinate(2, 3), new Cat(Owner.Player2));
+		catLoverPieces.put(new Coordinate(3, 2), new Cat(Owner.Player2));
+		catLoverPieces.put(new Coordinate(3, 3), new Cat(Owner.Player2));
+
+		catLoverPieces.put(new Coordinate(7, 7), new Rabbit(Owner.Player2));
+		catLoverPieces.put(new Coordinate(3, 6), new Rabbit(Owner.Player2));
+		catLoverPieces.put(new Coordinate(3, 4), new Camel(Owner.Player2));
+		catLoverPieces.put(new Coordinate(4, 4), new Elephant(Owner.Player2));
+		catLoverPieces.put(new Coordinate(0, 0), new Horse(Owner.Player2));
+		this.catLoverAi = new Ai(Owner.Player2, new Game(new BoardState(catLoverPieces)));
+	}
+
+	public void randomStressTest(HashMap<Object, Double> expectedPercentages, Generater method) {
+		HashMap<Object, Double> countMap = new HashMap<Object, Double>();
+		for (int i = 0; i < ITERATION_SIZE; i++) {
+			Object returnValue = method.generate();
+			if (!countMap.containsKey(returnValue)) {
+				countMap.put(returnValue, 0.0);
+			}
+			countMap.put(returnValue, countMap.get(returnValue) + 1);
+		}
+
+		// convert count to percentages
+		for (Object obj : countMap.keySet()) {
+			countMap.put(obj, round(countMap.get(obj) / ITERATION_SIZE, 2));
+		}
+
+		for (Object obj : countMap.keySet()) {
+			double expectedPercent = expectedPercentages.get(obj);
+			double expectedLow = Math.max(expectedPercent - RANDOM_MARGIN, Math.min(Double.MIN_NORMAL, expectedPercent));
+			double expectedHigh = Math.min(expectedPercent + RANDOM_MARGIN, Math.max(100.0 - Double.MIN_NORMAL, expectedPercent));
+			double percent = countMap.get(obj);
+			String errorString = obj.toString() + " was outside of the expected range of " + (expectedPercent * 100)
+					+ "% +- " + (RANDOM_MARGIN * 100) + "% with a percentage of " + (percent * 100) + "%";
+			if (expectedLow > percent || percent > expectedHigh) {
+				StackTraceElement methodName = Thread.currentThread().getStackTrace()[2];
+				System.err.println("FAIL " + methodName);
+				// System.err.println(countMap);
```

Comment:
to print out, when push and pulls, equals and hashcode functions are implemented